guardrails/validators.py

```diff
@@ -2229,3 +2229,140 @@ def query_vector_collection(
             raise ValueError("distance_metric must be 'cosine'.")
 
         return top_chunks
+
+
+@register_validator(name="distribution-check", data_type="string")
+class SimilarToList(Validator):
+    """Validates that a value is similar to a list of previously known values.
+
+    For integer values, this validator checks whether the value lies
+    within 'k' standard deviations of the mean of the previous values.
+    (Assumes that the previous values are normally distributed.)
+
+    For string values, this validator checks whether the average
+    semantic similarity between the generated value and the previous
+    values is less than a threshold.
+    """
+
+    def __init__(
+        self,
+        k: int = 3,
+        threshold: float = 0.1,
+        on_fail: Optional[Callable] = None,
+        **kwargs,
+    ):
+        """
+        Args:
+            k (int): The number of standard deviations from the mean to check.
+                Defaults to 3.
+            threshold (float): The threshold for the average semantic similarity.
+                Defaults to 0.1.
+        """
+        super().__init__(on_fail, k=k, threshold=threshold, **kwargs)
+        self._k = int(k)
+        self._threshold = float(threshold)
+
+    def _get_semantic_similarity(
+        self, text1: str, text2: str, embed_function: Callable
+    ) -> float:
+        """Get the semantic similarity between two strings.
+
+        Args:
+            text1 (str): The first string.
+            text2 (str): The second string.
+            embed_function (Callable): The embedding function.
+
+        Returns:
+            similarity (float): The semantic similarity between the two strings.
+        """
+        text1_embedding = embed_function(text1)
+        text2_embedding = embed_function(text2)
+
+        similarity = 1 - (
+            np.dot(text1_embedding, text2_embedding)
+            / (np.linalg.norm(text1_embedding) * np.linalg.norm(text2_embedding))
+        )
+        return similarity
+
+    def validate(self, value: Any, metadata: Dict) -> ValidationResult:
+        prev_values = metadata.get("prev_values", [])
+
+        if not prev_values:
+            raise ValueError("You must provide a list of previous values in metadata.")
+
+        # Check if np is installed
+        if not _HAS_NUMPY:
```

Comment:
Yeah, something along these lines:

```python
        try:
            import numpy  # type: ignore
        except ImportError:
            raise RuntimeError(
                "The `numpy` package is not installed. "
            )
```

Not sure what type of error we should raise, but that's the overall structure

Reply:
@zsimjee The only issue with this style would be - we would need to add this block in each function where `np` is used. This is because, to use any numpy functions it needs to be within the scope of the function.

In this validator, it is used in `get_semantic_similarity()` and `validate()`. Hence, right now, we have this try except block defined globally, towards the start of the program from lines 31-36, and we need to check whether `numpy` is installed with the variable `_HAS_NUMPY` only once per validator, instead of once per each function in each validator.

Reply:
> we would need to add this block in each function where np is used

That's correct, unless it's okay to throw the error on module import.
Try rebasing on `guardrails-internal`, and see if it satisfies pyright with `make type`

Reply:
Ok let's merge this first - we make these changes when strict typing is being merged in